x-pack/plugins/fleet/server/services/preconfiguration/outputs.ts

```diff
@@ -99,25 +108,90 @@ export async function createOrUpdatePreconfiguredOutputs(
       }
 
       const isUpdateWithNewData =
-        existingOutput && isPreconfiguredOutputDifferentFromCurrent(existingOutput, data);
-
-      if (isCreate) {
-        logger.debug(`Creating output ${output.id}`);
-        await outputService.create(soClient, esClient, data, { id, fromPreconfiguration: true });
-      } else if (isUpdateWithNewData) {
-        logger.debug(`Updating output ${output.id}`);
-        await outputService.update(soClient, esClient, id, data, { fromPreconfiguration: true });
-        // Bump revision of all policies using that output
-        if (outputData.is_default || outputData.is_default_monitoring) {
-          await agentPolicyService.bumpAllAgentPolicies(soClient, esClient);
-        } else {
-          await agentPolicyService.bumpAllAgentPoliciesForOutput(soClient, esClient, id);
+        existingOutput && (await isPreconfiguredOutputDifferentFromCurrent(existingOutput, data));
+
+      if (isCreate || isUpdateWithNewData) {
+        const secretHashes = await hashSecrets(output);
+
+        if (isCreate) {
+          logger.debug(`Creating preconfigured output ${output.id}`);
+          await outputService.create(soClient, esClient, data, {
+            id,
+            fromPreconfiguration: true,
+            secretHashes,
+          });
+        } else if (isUpdateWithNewData) {
+          logger.debug(`Updating preconfigured output ${output.id}`);
+          await outputService.update(soClient, esClient, id, data, {
+            fromPreconfiguration: true,
+            secretHashes,
+          });
+          // Bump revision of all policies using that output
+          if (outputData.is_default || outputData.is_default_monitoring) {
+            await agentPolicyService.bumpAllAgentPolicies(soClient, esClient);
+          } else {
+            await agentPolicyService.bumpAllAgentPoliciesForOutput(soClient, esClient, id);
+          }
         }
       }
     })
   );
 }
 
+export async function hashSecret(secret: string) {
+  return new Promise((resolve, reject) => {
+    const salt = crypto.randomBytes(16).toString('hex');
+    crypto.scrypt(secret, salt, 64, { p: 5 }, (err, derivedKey) => {
```

Comment:
nit: should the integer values be moved to constants?

Reply:
I found a minor improvement to make this more secure. If no objections, I'll move the hashing constants and functions to https://github.com/elastic/kibana/blob/main/x-pack/plugins/fleet/server/services/preconfiguration/utils.ts.